### PR TITLE
[MIG][12.0] port pos_order_mgmt

### PR DESCRIPTION
**Pull request by PierrickBrun**
With modules split as discussed in #334

### CODE REVIEW

**Comment by pedrobaeza** [comment]
A migration script will be needed that installs the 2 extra modules.

File: pos_order_mgmt/readme/USAGE.rst
Comment (legalsylvain) [inline pos_order_mgmt/readme/USAGE.rst]: replace by 
``.. image:: ../static/description/order-mgmt-icon.png``
With that syntax, 

- it will work locally in github interface if you see the USAGE.rst file in your browser
- and the main readme of the module will be correctly generated (ref : https://github.com/OCA/maintainer-tools/pull/397)

**Comment by PierrickBrun** [comment]
after trying to split it into several modules, it appeared to be more time consuming than I first thought.

I've migrated the module without splitting but I modified the pos.config to allow to choose to allow reprint or returning done orders.

I think the returning part is working well but I' don't know the 11.0 module and pos_order_return enough to be sure about it.

File: pos_order_mgmt/static/src/js/widgets.js
Comment (legalsylvain) [inline pos_order_mgmt/static/src/js/widgets.js]: why adding a double check, if the button is invisible ?

**Review by legalsylvain — COMMENTED** [review]
> after trying to split it into several modules, it appeared to be more time consuming than I first thought.

> I've migrated the module without splitting but I modified the pos.config to allow to choose to allow reprint or returning done orders.

Thanks @PierrickBrun for porting this module.

> I think the returning part is working well but I' don't know the 11.0 module and pos_order_return enough to be sure about it.

Well, ``pos_order_return`` was developed to 
- link returned order to the original order, (what your code does)
- to block the possibility  to return more quantity as the original one. This feature was developped only in backoffice part and had to be implemented in the front office.
As in new version of Odoo PoS, it is not possible to make an order in a backoffice part, we could consider ``pos_order_return`` as fully obsolete on the second point.

2. Did you checked if the module is compatible with ``pos_restaurant`` ? ref : https://github.com/OCA/pos/issues/333. CC @chienandalu, @carlosDomatix 

kind regards.

other wise, LGTM. (code review / no test).

**Comment by PierrickBrun** [comment]
Looks like it works better with pos_restaurant now

**Review by fuentes73 — APPROVED** [review]
Tested and working OK

File: pos_order_mgmt/static/src/js/widgets.js
Comment (fuentes73) [inline pos_order_mgmt/static/src/js/widgets.js:303]: Is missing at the beginning `var self = this;`
So this line throws the error: `Cannot read property 'db' of undefined`

File: pos_order_mgmt/static/src/js/widgets.js
Comment (fuentes73) [inline pos_order_mgmt/static/src/js/widgets.js]: Testing more in depth the module I also seen that, if you return and order and reload the page without change anything, the "return" info is lost.
return-pos-order


A possible solution might be to add `order.trigger('change');` before return the order
```suggestion
                order.trigger('change');
                return order;
```

**Comment by PierrickBrun** [comment]
thank you for the review @fuentes010 , I've addressed the changes

File: pos_order_mgmt/static/src/js/widgets.js
Comment (Kerrim-) [inline pos_order_mgmt/static/src/js/widgets.js]: Hi,

I'm using this code for a project and wanted to ask if it is possible to change this part to be a little easier to extend. For example by putting it in some kind of dispatch method or like this:

```suggestion
                if(correct_order_print){
                    self["action_" + action](order);
                }
```

**Pros**: Easier to extend (You only need to register the event and write the method)
**Cons**: If you need different checks before executing the action or you want to use actions not starting with action this suggestion won't work. (But the dispatch method still would.)

Comment (Kerrim-) [reply pos_order_mgmt/static/src/js/widgets.js]: Also if there is more work to do with this module I would be glad to help :)

**Comment by pedrobaeza** [comment]
Please add #354 here

File: pos_order_mgmt/static/src/js/widgets.js
Comment (PierrickBrun) [reply pos_order_mgmt/static/src/js/widgets.js]: Hi @Kerrim- , No problem about accepting a PR from my side :+1: I only use the module for reprinting tickets so I did not dig a lot into returns

**Comment by ivantodorovich** [comment]
Functional test only.

I only encounter this issue (very small one):
On 11.0. the order list window closes automatically when clicking the refund order button.

**12.0:**
9Lx4sxJ4Rw


**11.0:**
H4VJPthjQu

**Comment by ivantodorovich** [comment]
It's probably because of this line:
https://github.com/OCA/pos/pull/337/commits/e548fccd518bf2653723a1107d058b50a41dcdb5?file-filters%5B%5D=.js&owned-by%5B%5D=#diff-3619c2549c163cd99597081619ddb2c4R273

**Comment by PierrickBrun** [comment]
@ivantodorovich nice catch :+1:

**Review by ivantodorovich — APPROVED** [review]

**Comment by chienandalu** [comment]
Hi, @PierrickBrun , maybe you want to take a look at this: https://github.com/OCA/pos/pull/361

File: pos_order_mgmt/models/pos_order.py
Comment (legalsylvain) [inline pos_order_mgmt/models/pos_order.py:29]: Where is this function defined ?

File: pos_order_mgmt/static/src/js/widgets.js
Comment (legalsylvain) [inline pos_order_mgmt/static/src/js/widgets.js]: I don't understand in which case dataset.orderId is not defined, and order is not defined anywhere.

**Review by legalsylvain — CHANGES_REQUESTED** [review]
Hi @PierrickBrun. 

* some minor errors inline. 
* there are also a lot of lint warning on Travis. 
* Coverage is low because there is no python test
* Adding a ribbon is cool, but generate a display bug in the total place : 
image




I propose to fix all that points this week, in a PR against your branch.

Otherwise, thanks a lot for this work ! 
kind regards.

File: pos_order_mgmt/models/pos_config.py
Comment (legalsylvain) [inline pos_order_mgmt/models/pos_config.py]: This field is not used anywhere.

File: pos_order_mgmt/models/pos_config.py
Comment (legalsylvain) [inline pos_order_mgmt/models/pos_config.py]: there are two field with the same oldname value. 'iface_load_done_order'. why ?

**Comment by legalsylvain** [comment]
Hello, I finished the port of this module, fixing some points, and adding test. (coverage from 43% to 71%).
I did a PR against this PR : https://github.com/OCA/pos/pull/373
You can test it on the runbot of the PR.

regards.

File: pos_order_mgmt/models/pos_config.py
Comment (PierrickBrun) [reply pos_order_mgmt/models/pos_config.py]: the field iface_load_done_order is splitted, so if you used it, you want both the functions once migrated

**Review by legalsylvain — APPROVED** [review]

**Comment by legalsylvain** [comment]
@fuentes010, @ivantodorovich : various changes since your last review. do you want to update your review ? 
thanks.

**Review by rafaelbn — CHANGES_REQUESTED** [review]
Please don't merge until @chienandalu  make last review tomorrow, thanks!

**Review by chienandalu — APPROVED** [review]
Tested on runbot and code review. Thanks for the good work!

**Comment by legalsylvain** [comment]
Hi @chienandalu, thanks for your review.

/ocabot merge

**Pull request closed without merging**
